### ISSUE TITLE
An option for IntelliSense in code-fenced examples

### ISSUE THREAD

**Issue by kizu** [mdx-js/mdx-analyzer#323]
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

Provide IntelliSense for the _code examples_ in`.mdx`

We're using a pattern, where we place our code examples in code-fenced blocks like this:

````mdx
import { Button } from '…';

## Example

Here is a button:

```jsx
<Button label="I am a button" />
```
````

Then, we're using [react-live](https://github.com/FormidableLabs/react-live) to render those examples as live playgrounds.

The issue is that while mdx-analyzer handles the regular code in the MDX (awesome!), it does not treat the code inside the code-fenced blocks as actual code.

### Solution

What we would like to see — an _option_ (as I don't this is something that should be available by default, as not everyone is using this pattern of doing things) to treat certain code-fenced blocks as actual code.

### Alternatives

Not that I know of. With the requirement that we need to have the examples' source code available, we cannot just use a component wrapper around the code we want — we need the code-fenced blocks specifically to have the access to the actual code, with all the comments etc.

In addition to this, having the example code in code-fenced blocks makes authoring and reviewing markdown changes much better — in various editors and on GitHub we can see the code examples with proper markup, rather than as jsx content.

**Comment by wooorm** [comment]
Hmm, this feels a bit weird. The point of using code is to *not* evaluate things

How are you evaluating the code inside those code blocks?

---

> in various editors and on GitHub we can see the code examples with proper markup, rather than as jsx content.

Use this project or for other editors use https://github.com/wooorm/markdown-tm-language. Soon also in GH. So I don’t think that’s a solid reason.

**Comment by kizu** [comment]
> How are you evaluating the code inside those code blocks?

We're gathering the imports (and exported consts) from the .mdx, then provide them as the evaluation context to react-live (alongside some additional internal hooks that can be used in the example but which we don't want to import in _every_ .mdx).

The main idea for those code examples — they're their source code first (not evaluated), then as a progressive enhancement they're evaluated and shown _alongside_ an editable source, where we still want to keep all the comments, original formatting etc.

So, if we'd have to do something like 

```jsx
<Example>
    <Button label="I am a button" />
</Example>
```

instead, we would lose the ability to get the source code without some very serious hacks (like, is it even possible from the mdxast point of view get the _source_ of anything that is nested inside the `Example`?)

**Comment by kizu** [comment]
Thanks, I'll investigate this, _however_ I would say that ability to have IntelliSense inside regular code examples would still be helpful — like, we'd want the users of the docs to be sure that the docs are up-to-date, so if we could improve the examples in any way — it would be awesome.

**Comment by wooorm** [comment]
> we'd want the users of the docs to be sure that the docs are up-to-date

How do you mean? How doesn’t my above solution not do that?
What is an example of out-of-date code that isn’t up-to-date?

Or are you saying that all ` ```js/jsx/ts/tsx/md/mdx ` code blocks should be (optionally) handled by this project?

**Comment by remcohaszing** [comment]
I think #287 comes very close to what you want in a generic matter.

Your code requires some domain specific logic though. What you might want is to use a custom remark plugin for this. This is a more complex feature and is tracked in #297.

**Comment by remcohaszing** [comment]
I’ve closed this in favor of #287 and #297.
